rust/agama-network/src/model.rs

```diff
@@ -0,0 +1,439 @@
+        };
+        let conn0 = Connection::Ethernet(EthernetConnection { base });
+        state.add_connection(conn0).unwrap();
+        let found = state.get_connection(uuid).unwrap();
+        assert_eq!(found.uuid(), uuid);
+    }
+
+    #[test]
+    fn test_add_duplicated_connection() {
+        let mut state = NetworkState::default();
+        let uuid = Uuid::new_v4();
+        let base = BaseConnection {
+            uuid,
+            ..Default::default()
+        };
+        let conn0 = Connection::Ethernet(EthernetConnection { base });
+        state.add_connection(conn0.clone()).unwrap();
+        let error = state.add_connection(conn0).unwrap_err();
+        assert!(matches!(error, NetworkStateError::ConnectionExists(_)));
+    }
+
+    #[test]
+    fn test_update_connection() {
+        let mut state = NetworkState::default();
+        let uuid = Uuid::new_v4();
+        let base0 = BaseConnection {
+            uuid,
+            ..Default::default()
+        };
+        let conn0 = Connection::Ethernet(EthernetConnection { base: base0 });
+        state.add_connection(conn0).unwrap();
+
+        let base1 = BaseConnection {
+            uuid,
+            ..Default::default()
+        };
+        let conn2 = Connection::Ethernet(EthernetConnection { base: base1 });
+        state.update_connection(conn2).unwrap();
+        let found = state.get_connection(uuid).unwrap();
+        assert_eq!(found.uuid(), uuid);
+    }
+
+    #[test]
+    fn test_update_unknown_connection() {
+        let mut state = NetworkState::default();
+        let uuid = Uuid::new_v4();
+        let base = BaseConnection {
+            uuid,
+            ..Default::default()
+        };
+        let conn0 = Connection::Ethernet(EthernetConnection { base });
+        let error = state.update_connection(conn0).unwrap_err();
+        assert!(matches!(error, NetworkStateError::UnknownConnection(_)));
+    }
+
+    #[test]
+    fn test_remove_connection() {
+        let mut state = NetworkState::default();
+        let uuid = Uuid::new_v4();
+        let base0 = BaseConnection {
+            uuid,
+            ..Default::default()
+        };
+        let conn0 = Connection::Ethernet(EthernetConnection { base: base0 });
+        state.add_connection(conn0).unwrap();
+        state.remove_connection(uuid).unwrap();
+        let found = state.get_connection(uuid).unwrap();
+        assert!(found.is_missing());
+    }
+
+    #[test]
+    fn test_remove_unknown_connection() {
+        let mut state = NetworkState::default();
+        let error = state.remove_connection(Uuid::new_v4()).unwrap_err();
+        assert!(matches!(error, NetworkStateError::UnknownConnection(_)));
+    }
+}
+
+/// Network device
+#[derive(Debug, Clone)]
+pub struct Device {
+    pub name: String,
+    pub ty: DeviceType,
+}
+
+#[derive(Debug, PartialEq, Copy, Clone)]
+pub enum DeviceType {
+    Loopback = 0,
+    Ethernet = 1,
+    Wireless = 2,
+    Unknown = 3,
```

Comment:
I do not like this ordering as it means that if we add new type, it result in change of unknown value or having unknown as 3 forever with other having higher number. So I check https://dbus.freedesktop.org/doc/dbus-api-design.html and it also shows having unknown as 0 and not high number. So I suggest to have Unknown 0 and all known ones as +1